File: C++/sliding-window/3.longest-substring-without-repeating-characters.cpp
/*
 * @lc app=leetcode id=3 lang=cpp
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        int n = s.length();
        int maxLength = 0;
        int l = 0;
        unordered_set<char> charSet;

        for (char c : s)
        {
            while (charSet.count(c) > 0)
            {
                charSet.erase(s[l]);
                l += 1;
            }

            charSet.insert(c);

            maxLength = max(maxLength, (int)charSet.size());
        }

        return maxLength;
    }
};
// @lc code=end

File: Go/sliding-window/3.longest-substring-without-repeating-characters.go
/*
 * @lc app=leetcode id=3 lang=golang
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
func lengthOfLongestSubstring(s string) int {
	charSet := make(map[byte]bool)
	l := 0
	maxLength := 0

	for r, _ := range s {
		for charSet[s[r]] {
			delete(charSet, s[l])
			l++
		}

		charSet[s[r]] = true

		if r-l+1 > maxLength {
			maxLength = r - l + 1
		}
	}

	return maxLength
}

// @lc code=end

